src/middlewared/middlewared/plugins/directoryservices_/activedirectory_join_mixin.py

```diff
@@ -383,13 +398,21 @@ def _ad_join(self, job: Job, ds_type: DSType, domain: str):
 
         # remove stub krb5.conf to allow overriding with fix on KDC
         os.remove('/etc/krb5.conf')
-        self.middleware.call_sync('kerberos.do_kinit', {
-            'krb5_cred': cred,
-            'kinit-options': {
-                'kdc_override': {'domain': ad_config['domainname'], 'kdc': domain_info['kdc_server']}
-            }
-        })
-        self.middleware.call_sync('kerberos.wait_for_renewal')
-        self.middleware.call_sync('etc.generate', 'kerberos')
+        try:
+            self.middleware.call_sync('kerberos.do_kinit', {
+                'krb5_cred': cred,
+                'kinit-options': {
+                    'kdc_override': {'domain': ad_config['domainname'], 'kdc': domain_info['kdc_server']}
+                }
+            })
+        except KRB5Error:
+            # Attempt to kinit against DC we were talking to failed and so we'll switch to generic
+            # kinit loop to wait for things to settle.
+            self.logger.debug('Initial attempt to kinit with new kerberos pricipal failed. Starting kinit loop.', exc_info=True)
+            job.set_progress(80, 'Waiting for active directory to replicate machine account changes.')
+            self._ad_wait_kerberos_start()
+        else:
+            self.middleware.call_sync('kerberos.wait_for_renewal')
+            self.middleware.call_sync('etc.generate', 'kerberos')
```

Comment:
Just wondering.  Should we set the job progress to 100 at the end of this function? (95 before the service update?)

Reply:
At this point no, because the job is being wrapped around in 24.10. In future refactor work for 25.04, yes.